### PR TITLE
Add join method in str

### PR DESCRIPTION
**Pull request by Agent-Hellboy**

### CODE REVIEW

File: integration_tests/test_str_01.py [hunk 1]
```diff
@@ -45,6 +45,14 @@ def test_str_repeat():
     assert 3*a*3 == "XyzXyzXyzXyzXyzXyzXyzXyzXyz"
     assert a*-1 == ""
 
+def test_str_join():
+    a: str
+    a = ","
+    p:list[str] = ["a","b"]
+    res:str = a.join(p)
+    assert res == "a,b"
```
Comment (certik) [inline integration_tests/test_str_01.py:53]: Very nice. Let's also add a test where "a" is more than one character, say: `a = ", "` and `assert rest == "a, b"`.

**Review by certik — APPROVED** [review]
I think this is great, thank you @Agent-Hellboy. I left a minor suggestion.

**Comment by certik** [comment]
You also need to update our reference tests using: `./run_tests -u` and commit the changes.

File: integration_tests/test_str_01.py (hunk 1 above)
Comment (Agent-Hellboy) [reply integration_tests/test_str_01.py:53]: Actually it doesn't work for `",".join(["a,""b"])`.
 I need to make some changes, I have added a PR as a part of pycon India, I will continue after coming back to my hometown.

File: src/runtime/lpython_builtin.py [hunk 2]
```diff
@@ -764,6 +764,14 @@ def _lpython_str_upper(x: str) -> str:
     return res
 
 
+def _lpython_str_join(s:str, lis:list[str]) -> str:
+    res:str = ""
+    i:i32
+    for i in range(len(lis)):
```
Comment (OculusMode) [inline src/runtime/lpython_builtin.py:770]: Can be directly iterated on the list itself, this way it can be worked on iterables easily. Every iterable can be called using `next` so I think that would be more better.

something like this can be more cleaner I think.

```
for ele in range(list_):
    res+=ele
```

File: integration_tests/test_str_01.py (hunk 1 above)
Comment (certik) [reply integration_tests/test_str_01.py:53]: I see. We can still merge it as a start, then we can improve it.

File: src/runtime/lpython_builtin.py (hunk 2 above)
Comment (certik) [reply src/runtime/lpython_builtin.py:770]: If it works, we can use the simpler syntax. LPython doesn't yet work for user defined iterables, but eventually it will I think.

File: src/runtime/lpython_builtin.py [hunk 3]
```diff
@@ -764,6 +764,14 @@ def _lpython_str_upper(x: str) -> str:
...
+        res+=s
+        res+=lis[i]
+    return res[1:]
```
Comment (OculusMode) [inline src/runtime/lpython_builtin.py:773]: thats not reliable. You are always assuming your str len will be 1 here, It should fail when str has len more than 1. You can prob use `res[len(str):]`

Comment (certik) [reply src/runtime/lpython_builtin.py:773]: Good catch! I think even better would be something like:
```
if len(lis) == 0: return ""
res = lis[0]
for i in range(1, len(lis)):
    res += s + lis[i]
return res
```

Comment (OculusMode) [reply src/runtime/lpython_builtin.py:773]: That works too, altho as long as there plans of iterables, part of `len` will probably be removed regardless.

File: integration_tests/test_str_01.py (hunk 1 above)
Comment (Agent-Hellboy) [reply integration_tests/test_str_01.py:53]: sure, I have added a new test.

**Comment by Agent-Hellboy** [comment]
> You also need to update our reference tests using: `./run_tests -u` and commit the changes.

okay, the CI tests are failing because of this, I didn't understand `update our reference tests`, what does `run_tests -u` do? 

i don't see `-u` flag in run_tests 
```
    parser.add_argument("-j", "-n", "--no_of_threads", type=int,
                help="Parallel testing on given number of threads")
    parser.add_argument("-b", "--backends", nargs="*", default=["llvm", "cpython"],
                type=str, help="Test the requested backends (%s), default: llvm, cpython" % \
                        ", ".join(SUPPORTED_BACKENDS))
    parser.add_argument("-f", "--fast", action='store_true',
                help="Run supported tests with --fast")
    return parser.parse_args()
 ```

**Comment by Agent-Hellboy** [comment]
I guess, I need to parse over asr nodes of the list argument passed inside the join method to check if only str is passed or not 
https://github.com/lcompilers/lpython/pull/2354/files#diff-147457a9117ab6d023c02f9a359d6060c248c4216cfcb47b1fafd17d885063feR6672
I am still exploring the codebase, I will try to do it in coming days

**Comment by certik** [comment]
It's the one in the root directory:
```console
(lp) repos/lpython(main) $ ./run_tests.py --help
usage: run_tests.py [-h] [-u] [-l] [-t [TEST ...]] [-b [BACKEND ...]] [-v]
                    [--exclude-test [TEST ...]]
                    [--exclude-backend [BACKEND ...]] [--no-llvm]
                    [--skip-run-with-dbg] [-s] [--no-color]

LPython Test Suite

options:
  -h, --help            show this help message and exit
  -u, --update          update all reference results
...
```

**Comment by certik** [comment]
> I guess, I need to parse over asr nodes of the list argument passed inside the join method to check if only str is passed or not

That's needed for error messages. But first let's get the two existing tests that you added (they both look great) to pass all CI. After CI passes, then we can either merge, or you can add some tests for errors.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 4]
```diff
@@ -33,6 +33,8 @@
 #include <lpython/semantics/python_intrinsic_eval.h>
 #include <lpython/parser/parser.h>
 #include <libasr/serialization.h>
+#include <lpython/pickle.cpp>
+
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:37]: I think this might be the reason for the CI failure!
Remove this and see if it fixes the failure?
```suggestion
```

Comment (Agent-Hellboy) [reply src/lpython/semantics/python_ast_to_asr.cpp]: Thanks, CI checks paased

File: integration_tests/test_str_01.py [hunk 5]
```diff
@@ -55,6 +70,7 @@ def check():
     test_str_index()
     test_str_slice()
     test_str_repeat()
+    test_str_join()
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:73]: Let's test `test_str_join2`

**Review by Thirumalai-Shaktivel — APPROVED** [review]
LGTM, Thank you!

File: integration_tests/test_str_01.py [hunk 6]
```diff
@@ -45,6 +45,21 @@ def test_str_repeat():
...
+    p:list[str] = ["a","b"]
+    res:str = a.join(p)
+    assert res == "a,b"
+
+def test_str_join2():
+    a: str
+    a = "**"
+    p:list[str] = ["a","b"]
+    res:str = a.join(p)
+    assert res == "a**b"
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:60]: We can test more cases like empty strings, more items in the list (more than 3), and so on...
This can be done in the subsequent PRs.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 4 above)
Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp]: For reference, the CI failure was:
```
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle(struct LCompilers::ASR::TranslationUnit_t &,bool,bool,bool)" (?pickle@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUTranslationUnit_t@ASR@2@_N11@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle(struct LCompilers::ASR::asr_t &,bool,bool,bool)" (?pickle@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUasr_t@ASR@2@_N11@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_json(struct LCompilers::ASR::TranslationUnit_t &,struct LCompilers::LocationManager &,bool)" (?pickle_json@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUTranslationUnit_t@ASR@2@AEAULocationManager@2@_N@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_json(struct LCompilers::ASR::asr_t &,struct LCompilers::LocationManager &,bool)" (?pickle_json@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUasr_t@ASR@2@AEAULocationManager@2@_N@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_json(struct LCompilers::LPython::AST::ast_t &,struct LCompilers::LocationManager &)" (?pickle_json@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUast_t@AST@12@AEAULocationManager@2@@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_python(struct LCompilers::LPython::AST::ast_t &,bool,bool)" (?pickle_python@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUast_t@AST@12@_N1@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_tree(struct LCompilers::ASR::TranslationUnit_t &,bool,bool)" (?pickle_tree@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUTranslationUnit_t@ASR@2@_N1@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_tree(struct LCompilers::ASR::asr_t &,bool,bool)" (?pickle_tree@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUasr_t@ASR@2@_N1@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
lpython_lib.lib(python_ast_to_asr.cpp.obj) : error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl LCompilers::LPython::pickle_tree_python(struct LCompilers::LPython::AST::ast_t &,bool)" (?pickle_tree_python@LPython@LCompilers@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAUast_t@AST@12@_N@Z) already defined in lpython_lib.lib(pickle.cpp.obj)
   Creating library src\bin\lpython.lib and object src\bin\lpython.exp

src\bin\lpython.exe : fatal error LNK1169: one or more multiply defined symbols found
```

File: integration_tests/test_str_01.py (hunk 1 above)
Comment (certik) [reply integration_tests/test_str_01.py:53]: I see it, thanks.

**Review by certik — APPROVED** [review]
I think this looks good, thanks!

**Comment by Shaikh-Ubaid** [comment]
I just noticed this. @Agent-Hellboy this is amazing! I really appreciate it. Keep up the great work.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 7]
```diff
@@ -6664,6 +6664,20 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                throw SemanticError("str.join() takes one argument",
+                    loc);
+            }
+            fn_call_name = "_lpython_str_join";
+            ASR::call_arg_t str_var;
+            str_var.loc = loc;
+            str_var.m_value = s_var;
+            ASR::call_arg_t passed_int;
+            passed_int.loc = loc;
+            passed_int.m_value = args[0].m_value;
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:6678]: I just noticed this. Here, I think `passed_int` should be named `list_of_str` (or something similar). (@Agent-Hellboy  I remember we were experimenting with it by passing an integer.)

I think you can fix the above in your new PR https://github.com/lcompilers/lpython/pull/2364. You can just make the rename a separate commit.

Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:6678]: Go ahead and create an issue, so that we don't forget.

Comment (Agent-Hellboy) [reply src/lpython/semantics/python_ast_to_asr.cpp:6678]: created https://github.com/lcompilers/lpython/issues/2365
